apps/emqx/src/emqx_persistent_session_ds.erl

```diff
@@ -433,24 +434,27 @@ handle_timeout(_ClientInfo, bump_last_alive_at, Session0) ->
 
 -spec replay(clientinfo(), [], session()) ->
     {ok, replies(), session()}.
-replay(_ClientInfo, [], Session = #{inflight := Inflight0}) ->
+replay(
+    ClientInfo,
+    [],
+    Session = #{inflight := Inflight0, subscriptions := Subs, props := Conf}
+) ->
+    UpgradeQoS = maps:get(upgrade_qos, Conf),
     AckedUntil = emqx_persistent_message_ds_replayer:committed_until(ack, Inflight0),
     RecUntil = emqx_persistent_message_ds_replayer:committed_until(rec, Inflight0),
     CompUntil = emqx_persistent_message_ds_replayer:committed_until(comp, Inflight0),
-    ReplyFun = fun
+    ReplyFun = make_reply_fun(ClientInfo, Subs, UpgradeQoS, fun
         (_Seqno, #message{qos = ?QOS_0}) ->
-            {false, []};
+            [];
         (Seqno, #message{qos = ?QOS_1}) when Seqno < AckedUntil ->
             [];
         (Seqno, #message{qos = ?QOS_2}) when Seqno < CompUntil ->
             [];
```

Comment:
Q: with the change in `replayer:publish`, these clauses will stop advancing `Seqno` in the accumulator in `replayer:publish` whereas before they did.  Is this the intention?

Reply:
Oh, right. 🫠 Very good catch!

Reply:
Should be fixed in e02e635.